core/src/main/java/hudson/model/DownloadService.java

```diff
@@ -195,6 +199,45 @@ public static String loadJSONHTML(URL src) throws IOException {
     }
 
     /**
+     * This installs itself as a listener to changes to the Downloadable extension list and will download the metadata
+     * for any newly added Downloadables.
+     */
+    @Restricted(NoExternalUse.class)
+    public static class DownloadableListener extends ExtensionListListener {
+
+        /**
+         * Install this listener to the Downloadable extension list after all extensions have been loaded; we only
+         * care about those that are added after initialization
+         */
+        @Initializer(after = InitMilestone.EXTENSIONS_AUGMENTED)
+        public static void installListener() {
+            ExtensionList.lookup(Downloadable.class).addListener(new DownloadableListener());
+        }
+
+        /**
+         * Look for Downloadables that have no data, and update them.
+         */
+        @Override
+        public void onChange() {
+            for (Downloadable d : Downloadable.all()) {
+                TextFile f = d.getDataFile();
+                if (f == null || !f.exists()) {
+                    LOGGER.log(Level.FINE, "Updating metadata for " + d.getId());
+                    try {
+                        d.updateNow();
+                    } catch (IOException e) {
```

Comment:
@oleg-nenashev The caller `ExtensionListListener` already handles these, so in the worst case not all necessary `Downloadable`s are refreshed. Since I gate this with a check to only refresh those without data, it should only happen for a `Downloadable` in a plugin with 2+ `Downloadable`s that is broken anyway.